modules/joinmarket.nix

```diff
@@ -235,27 +236,44 @@ in {
       requires = [ "bitcoind.service" ];
       after = [ "bitcoind.service" ];
       preStart = ''
-        install -o '${cfg.user}' -g '${cfg.group}' -m 640 ${configFile} ${cfg.dataDir}/joinmarket.cfg
-          sed -i \
-            "s|@@RPC_PASSWORD@@|rpc_password = $(cat ${secretsDir}/bitcoin-rpcpassword-privileged)|" \
-            '${cfg.dataDir}/joinmarket.cfg'
-        '';
+        {
+          cat ${configFile}
+          echo
+          echo '[BLOCKCHAIN]'
+          echo "rpc_password = $(cat ${secretsDir}/bitcoin-rpcpassword-privileged)"
+        } > '${cfg.dataDir}/joinmarket.cfg'
+      '';
       # Generating wallets (jmclient/wallet.py) is only supported for mainnet or testnet
       postStart = mkIf (bitcoind.network == "mainnet") ''
         walletname=wallet.jmdat
         wallet=${cfg.dataDir}/wallets/$walletname
         if [[ ! -f $wallet ]]; then
           ${optionalString (cfg.rpcWalletFile != null) ''
             echo "Create watch-only wallet ${cfg.rpcWalletFile}"
-            ${bitcoind.cli}/bin/bitcoin-cli -named createwallet \
-              wallet_name="${cfg.rpcWalletFile}" disable_private_keys=true
+            if ! output=$(${bitcoind.cli}/bin/bitcoin-cli -named createwallet \
+                          wallet_name="${cfg.rpcWalletFile}" disable_private_keys=true 2>&1); then
+              # Ignore error if bitcoind wallet already exists
+              if [[ $output != *"already exists"* ]]; then
+                echo "$output"
+                exit 1
+              fi
+            fi
           ''}
-          pw=$(cat "${secretsDir}"/jm-wallet-password)
+          # Restore wallet from seed if available
+          seed=
+          if [[ -e jm-wallet-seed ]]; then
+            seed="--recovery-seed-file jm-wallet-seed"
+          fi
           cd ${cfg.dataDir}
-          if ! ${nbPkgs.joinmarket}/bin/jm-genwallet --datadir=${cfg.dataDir} $walletname $pw \
-                 | grep 'recovery_seed' \
-                 | cut -d ':' -f2 \
-                 | (umask u=r,go=; cat > jm-wallet-seed); then
+          # Strip trailing newline from password file
+          if ! tr -d "\n" <"${secretsDir}/jm-wallet-password" \
+               | ${nbPkgs.joinmarket}/bin/jm-genwallet \
+                   --datadir=${cfg.dataDir} --wallet-password-stdin $seed $walletname \
+               | grep 'recovery_seed' \
+               | cut -d ':' -f2 \
```

Comment:
Getting

```
joinmarket-post-start[550448]: cut: write error: Broken pipe`
```

on both `--recovery-seed-file` and normal generation. Can't immediately identify where this is coming from.

Reply:
This error is caused by the subshell containing the `if` statement exiting early when the condition is false. This error never occured in my local testing or in CI.
Fixed [here](https://github.com/erikarvstedt/nix-bitcoin/commits/joinmarket-0.9.1-ea-2).